### ISSUE TITLE
Typescript Support?

### ISSUE THREAD

**Issue by keeslinp** [FreeMasen/RESP#33]
Hey, I love the idea behind this and RESSA. I'm hoping to use this in a project of mine and I'm mostly working with typescript. Would you be opposed to a PR adding typescript support? It might be a while until I get to that portion of the project but I figured I'd float it out there just in case you don't want to support typescript.

**Comment by allsey87** [comment]
Depending on how much work this is involved, I might be able to work on this a bit if we can also handle typescript definition files. This is what I have found so far by probing the code a bit.
```ts
export let SomeVariable: number;
```
the parser currently chokes in `parse_lexical_decl` when it encounters a column denoting the start of the type specification. Similarly, when a type specifier on a class field is encounted, e.g., `someField: SomeType;` the parser chokes in `parse_class_el`.

For a class function (like the following), the parser chokes on the column after  `someArgument` and if removed, again just before `SomeReturnType`
```ts
someMethod( someArgument: SomeType ): SomeReturnType;
```
For the former, parsing breaks in `parse_property_method` and for the latter, in one of the methods for parsing the method body.

So basically we need to extend the parser to handle the colon in these contexts. @FreeMasen do you have any recommendations on how I could proceed?

**Comment by allsey87** [comment]
These are a few tests that I have put together that currently fail
```rust
extern crate ressa;

#[test]
fn type_annotation_in_export() {
    let tsd = "
export let someNumber: number;
";
    let _ = ressa::Parser::builder()
        .module(true)
        .js(tsd)
        .build()
        .expect("failed to create parser")
        .parse()
        .expect("failed to parse tsd");
}

#[test]
fn type_annotation_in_field() {
    let tsd = "
export class SomeClass {
   someField : number;
};
";
    let _ = ressa::Parser::builder()
        .module(true)
        .js(tsd)
        .build()
        .expect("failed to create parser")
        .parse()
        .expect("failed to parse tsd");
}

#[test]
fn type_annotation_in_method_args() {
    let tsd = "
export class SomeClass {
   someMethod(someArgument: number);
};
";
    let _ = ressa::Parser::builder()
        .module(true)
        .js(tsd)
        .build()
        .expect("failed to create parser")
        .parse()
        .expect("failed to parse tsd");
}

#[test]
fn type_annotation_in_method_return() {
    let tsd = "
export class SomeClass {
   someMethod(): number;
};
";
    let _ = ressa::Parser::builder()
        .module(true)
        .js(tsd)
        .build()
        .expect("failed to create parser")
        .parse()
        .expect("failed to parse tsd");
}
```

**Comment by allsey87** [comment]
So I am willing to put several hours into getting typescript support into RESSA, but I do need some support or at least acknowledgement that this feature is still desirable. As the situation stands, I am currently having to look into using [swc](https://swc-project.github.io/) but I still have a strong preference to use/contribute to this project.

**Comment by FreeMasen** [comment]
@allsey87 Thank you for getting started on this. I would be happy to work with you to implement a typescript parsing option for this project! Sorry it has taken me so long to respond.

One of the big questions that I still don't feel like I have a good answer for is how do we output something that is meaningful for a typescript file but doesn't complicate the JavaScript user experience.  Currently the [AST for javascript](https://github.com/rusty-ecma/RESAST) isn't going to have the option of adding a "type ident" for expressions. 

 If we choose to extend the AST in place and add an optional `type_identifier` this would drastically change the AST. If we define a seperate TypeScript AST, we would need to maintain a second parser to that returns this new AST from `next`. 

I have a few other ideas about how this _could_ go but I haven't had a chance to really flesh them out. If you have an idea about how to solve this I'd love to hear it!

**Comment by allsey87** [comment]
These design issues can be tricky to solve and to be honest, coming from a background of microcontroller programming and C/C++, I only understand and use Javascript and Typescript at a very high level and I am not in a knowledgeable-enough position to be making any suggestions on this front. I was under the impression initially that Typescript would just involve adding an additional field for a type identifier against several node types in the AST (and that typescript could be just enabled as a "dialect"), however, having looked at swc a bit, I can see now that it is not so straightforward.
